### ISSUE TITLE
code organization/structure

### ISSUE THREAD

**Issue by sarbbottam** [dequelabs/axe-core#189]
IMHO, the current code organization/structure is confusing. One has to go through the `Gruntfile.js` to comprehend what's happening. The build process, relies heavily on file concatenation. Wondering, if this project would embrace `ES2015` modules or at least `common.js` pattern?

Let the `project.json` and the file references dictate control instead of file concatenation.

The project already uses `babel` why not let babel perform the build and for `umd` generation, either `webpack` or `browserify` should be good enough.

**Comment by WilcoFiers** [comment]
This is an excellent suggestion. The addition of Babel was pretty recent, so this hasn't been an option before. This is a substantial change, but I think a good one to make! Thanks for the suggestion.

**Comment by sarbbottam** [comment]
> This is a substantial change, but I think a good one to make! 

No doubt in it, I guess it has to be a parallel effort.

**Comment by marcysutton** [comment]
I've filed an issue for this internally so we can discuss it as we improve the codebase.

**Comment by sarbbottam** [comment]
Let me know if I can be of any help, I would love to do so.

**Comment by marcysutton** [comment]
We would absolutely love help. This change would be a bit of all-or-nothing, so we'll want to be sure it's done in a branch and only merged if tests are passing and it builds successfully.

@sarbbottam want to have a call in the next week or two to discuss? We can set expectations and estimate the level of effort.

**Comment by sarbbottam** [comment]
@marcysutton sounds good.

**Comment by marcysutton** [comment]
Hi @sarbbottam, checking back up on this: are you still interested in an aXe contribution? My email address is in my profile.

**Comment by sarbbottam** [comment]
Hi @marcysutton I am still interested, I guess you wanted to setup a call to discuss the approach.

**Comment by sarbbottam** [comment]
I have been trying to comprehend the code organization and started with build workflow. Looks like there is quite a bit of code generation going on at [`grunt:configure`](https://github.com/dequelabs/axe-core/blob/master/Gruntfile.js#L74-L85). I also feel that 'configure task' is mislabeled, behind the scene, it is generating the `tmp/rules.js` along with `doc/rule-descriptions.md`. I guess in the process of staying DRY, the code organization has become more abstract and highly dependent on configuration, file structure and custom code generation.

<img width="1440" alt="magic" src="https://cloud.githubusercontent.com/assets/949380/20115753/5c666636-a5ae-11e6-9310-b317c8a03b90.png">

I found it very difficult to follow, I have been trying to co-relate the `axe.js` and the source code and still struggling.

I also feel `grunt:validate` and `grunt:lint` should be together as they are similar functionality.

I believe `grunt:configure` should be replaced by proper code relation, common.js export/require or es-next export/import.  And if that can be achieved, `grunt:build` can be completely replaced with `webpack` or `browserify` for `umd` generation.

**Comment by dylanb** [comment]
The reason for the code generation is compelling and simple:

1. We want to separate the language dependent pieces of the rules from the code, and

1. We want to separate the configurable portions of the code from the code, and

1. We want to be able to re-use the code, not copy and not re-implement, and

1. We want to be able to add rules and checks at runtime, and

1. We want to be able to reconfigure the rules and checks at runtime, and

1. We want to be able to change the messages associated with the rules and checks at run time

There may be other reasons I am forgetting.

Any refactoring will have to support all of the above in order to be acceptable. Difficulty to understand the build code is not an acceptable reason for changing anything - except the build code.

So if you want to rename "configure" to some other verb that is more understandable, please go ahead and propose something.

Also: validate validates the JSON structures of the rules and checks, this is very different from the source code linting. It is not linting in the sense of "does the syntax conform to the language specification?", it is validation in the sense of "does this JSON structure have all the correct and required data elements in order for a built ruleset to function correctly?"

**Comment by sarbbottam** [comment]
>We want to separate the configurable portions of the code from the code, and
>We want to be able to re-use the code, not copy and not re-implement, and
>We want to be able to add rules and checks at runtime, and
>We want to be able to reconfigure the rules and checks at runtime, and
>We want to be able to change the messages associated with the rules and checks at run time

Alternating the code architecture must not affect the above use cases.

>We want to separate the language dependent pieces of the rules from the code, and

Please excuse my ignorance, I don't think this is a valid use case. The configuration needs to be passed as a JavaScript object. It is not language independent. 

>Difficulty to understand the build code is not an acceptable reason for changing anything - except the build code.

Difficulty to understand the code pushed me to understand how code is generated, not the otherwise.

>Also: validate validates the JSON structures of the rules and checks, this is very different from the source code linting. It is not linting in the sense of "does the syntax conform to the language specification?", it is validation in the sense of "does this JSON structure have all the correct and required data elements in order for a built ruleset to function correctly?"

At the top level both are performing static analysis of the code and if you will mark the build fail if the lint fails then let the build fail at the very beginning instead of the last step.

**Comment by sarbbottam** [comment]
I as a user of axe-core don't feel confident about contributing as I don't understand the code, thus opened up this issue. Feel free to close the issue if you think otherwise.

**Comment by dylanb** [comment]
I think we all agree that simplifying the code is a good thing. The only point I am making is that it is not the primary concern - the primary concern is that the code should be usable by the end-users.

We have many, many end users who need to manipulate the rules that run in many different ways (as pointed out above). That is the primary driver for the design.

I agree with you that the validation should happen at the beginning of the build.

The configuration allows for language independence of the results structure because you can translate the strings, pass the translated strings into the `.configure` and the end result is a results structure in a new human language. We would like to improve the translatability of the messages but at least for now, we have some ability to achieve this.

**Comment by marcysutton** [comment]
What we want to improve is the build process for combining of different parts of aXe into one JavaScript file. That should not impact the end-user API for adding and configuring rules and checks at runtime–either the default code is included in the build or a user will supply it when they run `axe.configure`. 

One thing I would like to see change is how we use Babel while concatenating files together–the two processes are not working well together. I had opened this issue internally but shelved it since on its own it didn't seem important enough to fix. But if we are changing the infrastructure, I'd like to address this:

> In axe.js and axe.min.js, there is a repeated block of code in all of the core files:
```javascript
 var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function(obj) {
    return typeof obj;
  } : function(obj) {
    return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj;
  };
```
> This is caused by the babel:core task, which is called prior to concat:engine. The code block is repeated 7 times, redefining var = _typeof over and over. It would be cool to clean this up so the axe.js file is more lean.

> We could either run Babel after concat:engine or adjust our concatenation scheme to work better with Babel.

**Comment by dylanb** [comment]
That would be a good improvement - but in the hierarchy of priorities, readability of generated code falls below ease of development

**Comment by sarbbottam** [comment]
>readability of generated code falls below ease of development

That's not even a priority, neither should be a use-case, IMHO.
**I was trying to read the generated code for reverse engineering, I could not comprehend the flow from the current source code organization.**

**Comment by WilcoFiers** [comment]
Closing the issue, as it has gone quite stale.
